checks/frozen_deps.go

```diff
@@ -35,19 +36,115 @@ func FrozenDeps(c *checker.CheckRequest) checker.CheckResult {
 		isPackageManagerLockFilePresent,
 		isGitHubActionsWorkflowPinned,
+		isDockerfilePinned,
 	)(c)
 }
 
-// TODO(laurent): need to support Docker https://github.com/ossf/scorecard/issues/403
 // TODO(laurent): need to support GCB
 
+// ============================================================
+// ======================== Dockerfiles =======================
+// ============================================================
+func isDockerfilePinned(c *checker.CheckRequest) checker.CheckResult {
+	return CheckFilesContent(frozenDepsStr, "*Dockerfile*", false, c, validateDockerfile)
+}
+
+func isPresent(slice []string, val string) bool {
+	for _, item := range slice {
+		if item == val {
+			return true
+		}
+	}
+	return false
+}
+
+func validateDockerfile(path string, content []byte,
+	logf func(s string, f ...interface{})) (bool, error) {
+	// Users may use various names, e.g.,
+	// Dockerfile.aarch64, Dockerfile.template, Dockerfile_template, dockerfile, Dockerfile-name.template
+	// Templates may trigger false positives, e.g. FROM { NAME }.
+
+	// We have what looks like a docker file.
+	// Let's interpret the content as utf8-encoded strings.
+	scanner := bufio.NewScanner(strings.NewReader(string(content)))
+	asRegex := regexp.MustCompile(`^(?i)FROM\s+(.*)\s+AS\s+(.*)`)
+	regex := regexp.MustCompile(`^(?i)FROM\s+(.*)`)
+	hashAsRegex := regexp.MustCompile(`^(?i)FROM\s+.*@sha256:[a-f\d]{64}\s+AS\s+(.*)`)
+	hashRegex := regexp.MustCompile(`^(?i)FROM\s+.*@sha256:[a-f\d]{64}`)
+
+	// Read the file line by line.
+	scanner.Split(bufio.ScanLines)
+	r := true
+	fromLineNb := 0
+	var pinnedAsNames []string
+	for scanner.Scan() {
+		line := scanner.Text()
+		// Only look at lines starting with FROM.
+		if !strings.HasPrefix(strings.ToLower(line), "from ") {
+			continue
+		}
+
+		// New line found
+		fromLineNb += 1
+
+		// FROM name@sha256:hash AS newname.
+		// In this case, we record newname. It's pinned
+		// so it can be re-used as 'FROM new name' later.
+		re := hashAsRegex.FindStringSubmatch(line)
+		if len(re) == 2 {
+			// Record the newname.
+			pinnedAsNames = append(pinnedAsNames, re[1])
+			continue
+		}
+
+		// FROM oldname AS newname
+		// where oldname refers to a pinned image
+		re = asRegex.FindStringSubmatch(line)
+		if len(re) == 3 {
+			oldname := re[1]
+			newname := re[2]
+			if !isPresent(pinnedAsNames, oldname) {
+				r = false
+				logf("!! frozen-deps - %v has non-pinned dependency '%v'", path, line)
+				continue
+			}
+			// Record the newname if not alresdy present in our list.
+			if !isPresent(pinnedAsNames, newname) {
+				pinnedAsNames = append(pinnedAsNames, newname)
+			}
+			continue
+		}
+
+		// FROM name
+		// where name refers to a pinned image
+		re = regex.FindStringSubmatch(line)
+		if len(re) == 2 && isPresent(pinnedAsNames, re[1]) {
+			continue
+		}
+
+		// FROM name@sha256:hash
+		if !hashRegex.Match([]byte(line)) {
+			r = false
+			logf("!! frozen-deps - %v has non-pinned dependency '%v'", path, line)
```

Comment:
Here and other place, can we do early error return instead of logging the failures?

Reply:
I did it this way so we can log all failures. wdut?

Reply:
I do think we should get details of all failures, but it probably requires a decent amount code re-write. Ideally, we should have a way to return a list of errors and we can pile all of these errors onto it. 

But that might be too much to attempt in this PR, so suggested simply early returning for now.